### ISSUE TITLE
Point layer API expension : static point coordinates with time varying property

### ISSUE THREAD

**Issue by EmeEmu** [napari/napari#2592]
## 🚀 Feature
Would it be possible to add the possibility to modify points' color when moving in one dimension ?

## Motivation
You can find all the relevant information in the following [image.sc post](https://forum.image.sc/t/3d-point-cloud-with-changing-color-in-time/51635), but here is a quick summary.

Let's say that I have a 3D point cloud array `[ point , xyz ]` containing `n_points`, and I want to change the color of each point in time (with `n_step` time steps). 

If `n_points` and `n_step` are small, I can simply copy the x,y,z locations of all the points for all time steps (see the [image.sc post](https://forum.image.sc/t/3d-point-cloud-with-changing-color-in-time/51635) for more details), but for large `n_points` and `n_step`, this requires a lot of memory.

@jni has found a very nice workaround using a callback on the time slider (see the [image.sc post](https://forum.image.sc/t/3d-point-cloud-with-changing-color-in-time/51635) and the PR #2582). This works pretty well, but I think that adding this feature directly in the API would be very useful, and possibly more efficient. 

What do you think ? 

Thanks a lot

EmeEmu

**Comment by tlambert03** [comment]
I’m supportive!

**Comment by jni** [comment]
Yeah we discussed this in the Eurasia meeting with @kevinyamauchi and others. It's probably quite doable to allow nD properties "columns" that we broadcast against the points, just like NumPy broadcasting. Then, if any property is more than 1D, we broadcast it against the points and put a slider corresponding to the columns. I think ultimately it would be very clean, consistent with our other dimensionality design choices, and indeed it would allow for another item on your Wishlist @EmeEmu, having the nD points without the need for a higher-dimensional image underlying them. 

I'm excited!

**Comment by sofroniewn** [comment]
I'm supportive too, we already allow something similar in the surface layer for timeseries where the values change with a fixed surface. I have some question about whether this broadcasting leads to extra sliders in the dims area or whether we want to think of those extra sliders in a properties focused widget that was its own dock widget. I'd probably lean towards the later.

I think a general properties widget could support behavior like this and would be widely usable for us - cc @kevinyamauchi with whom we've discussed that idea before too

**Comment by jni** [comment]
@sofroniewn I actually think they should be dims sliders, since for example the idea here would be to match the points changing over time to some 4D images.

**Comment by sofroniewn** [comment]
Maybe @jni can you provide an example of how you see the API?

This is what i'm imagining now, but correct me if you see different

```python
add_points(data_xyz, properties={'values': points_timeseries})
```

where `data_xyz` has shape `Nx3` and `points_timeseries` has shape `NxK` say and then you'd get 1 extra dims slider with `K` steps. If `points_timeseries` has shape `NxKxL` then you'd get 2 extra dims sliders with `K` and `M` steps respectively.

I guess I'm ok with this, but would it depend on `values` being the key in the properties dictionary that we allowed this for, what is someone was did `properties={'my_special_name': points_timeseries}` . I'm just a little worried about the potential for ballooning complexity

**Comment by jni** [comment]
@sofroniewn I see it as:

```python
add_points(
    data_xyz,
    properties={
        'values0': points_timeseries,
        'values1': points_static,
        'values2': points_timeseries2,
    },
)
```

Here, `data_xyz` is NxD, `points_timeseries` is TxN, and `points_static` is N. `points_timeseries2` can be TxN or SxTxN or Sx1xN — anything that broadcasts with N. The key restrictions are (1) that `all(prop.shape[-1] == N for prop in properties.values())` and (2) that `np.broadcast_arrays(*properties.values())` should not raise an error.

The sliders would appear when any of our point visual properties (face color, edge color, size, or text) are mapped to one of the >1D arrays. When there are multiple such arrays selected, we broadcast them, call the resulting dimension E, and we make E-1 sliders.

**Comment by jni** [comment]
Alternately (probably better), we can add the sliders at all times — this points layer has dimension `data_xyz.shape[1] + np.broadcast_arrays(*properties.values())[0].ndim - 1`. But depending on what's being displayed you may not see any changes when you move the slider.

**Comment by sofroniewn** [comment]
> The sliders would appear when any of our point visual properties (face color, edge color, size, or text) are mapped to one of the >1D arrays. When there are multiple such arrays selected, we broadcast them, call the resulting dimension E, and we make E-1 sliders.
> ...
> Alternately (probably better), we can add the sliders at all times

Ok, I can see pros/ cons of either. In general though it's now all bout slicing with points we want to see out of our big nD space (hence wanting to reuse dims sliders), and if we then make a "properties" widget then that is more about editing the properties.

**Comment by brisvag** [comment]
I love this! I think it's important to keep in mind that this may not be used with points only.

I can definitely imagine having a 3D+time image and points that have some static property and some other property that changes over time... In that case, I'd want the property to be controlled by the dims slices, for sure!
